### ISSUE TITLE
TF: Can't create layer "model_1/reshape_1/Shape" of type "Shape"

### ISSUE THREAD

**Issue by SXQ233** [opencv/opencv#22926]
### System Information

OpenCV => 4.2.0
Operating System / Platform => Windows 10 64 bit
Compiler => Visual Studio 2017
TensorFlow 2.2.0

### Detailed description

I have trained a segnet using tf.keras. Then I exported the model as frozen graph *.pb file, and tried to load the network using cv::dnn::readNetFromTensorflow(). When I run the app the following error occurs:
OpenCV(4.2.0) Error: Unspecified error (Can't create layer "model_1/reshape_1/Shape" of type "Shape") in cv::dnn::dnn4_v20191202::LayerData::getLayerInstance, file C:\build\master_winpack-build-win64-vc15\opencv\modules\dnn\src\dnn.cpp, line 562.

### Steps to reproduce

def _segnet(n_classes, encoder, input_height=416, input_width=416, encoder_level=3):

    img_input, levels = encoder(input_height=input_height, input_width=input_width)

    feat = levels[encoder_level]
    o = segnet_decoder(feat, n_classes, n_up=3)

    n_up = 3
    o = ZeroPadding2D((1, 1))(feat)
    o = Conv2D(512, (3, 3), padding='valid')(o)
    o = BatchNormalization()(o)

    o = UpSampling2D((2, 2))(o)
    o = ZeroPadding2D((1, 1))(o)
    o = Conv2D(256, (3, 3), padding='valid')(o)
    o = BatchNormalization()(o)

    for _ in range(n_up - 2):
        o = UpSampling2D((2, 2))(o)
        o = ZeroPadding2D((1, 1))(o)
        o = Conv2D(128, (3, 3), padding='valid')(o)
        o = BatchNormalization()(o)

    o = UpSampling2D((2, 2))(o)
    o = ZeroPadding2D((1, 1))(o)
    o = Conv2D(64, (3, 3), padding='valid')(o)
    o = BatchNormalization()(o)

    o = UpSampling2D((2, 2))(o)
    o = ZeroPadding2D((1, 1))(o)
    o = Conv2D(16, (3, 3), padding='valid',name='output_16')(o)
    o = BatchNormalization()(o)

    o = Conv2D(n_classes, (3, 3), padding='same',name='output_2')(o)
    
    o = Reshape((int(input_height) * int(input_width ), n_classes))(o)
    o = Softmax()(o)
    model = Model(img_input, o)

    return model

### Issue submission checklist

- [X] I report the issue, it's not a question
- [ ] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by asmorkalov** [comment]
OpenCV 4.2.0 is very old. There are a lot of fixes and features have been done since that time. Please try the latest release.

**Comment by SXQ233** [comment]
I updated to the OpenCV 4.6.0 and the issue is still there.

Can't parse layer for node='model_1/reshape_1/Reshape/shape' of type='Pack'. Exception: OpenCV(4.6.0) c:\build\master_winpack-build-win64-vc15\opencv\modules\dnn\src\tensorflow\tf_importer.cpp:2799: error: (-2:Unspecified error) Input layer not found: model_1/reshape_1/Reshape/shape/1 in function 'cv::dnn::dnn4_v20220524::`anonymous-namespace'::TFImporter::connect'

**Comment by SXQ233** [comment]
> OpenCV 4.2.0 is very old. There are a lot of fixes and features have been done since that time. Please try the latest release.

I updated to the OpenCV 4.6.0 and the issue is still there.

Can't parse layer for node='model_1/reshape_1/Reshape/shape' of type='Pack'. Exception: OpenCV(4.6.0) c:\build\master_winpack-build-win64-vc15\opencv\modules\dnn\src\tensorflow\tf_importer.cpp:2799: error: (-2:Unspecified error) Input layer not found: model_1/reshape_1/Reshape/shape/1 in function 'cv::dnn::dnn4_v20220524::`anonymous-namespace'::TFImporter::connect'

**Comment by zihaomu** [comment]
Hi @SXQ233, can you attach the failed `.pb` file to this issue?

**Comment by SXQ233** [comment]
> Hi @SXQ233, can you attach the failed `.pb` file to this issue?

**Comment by SXQ233** [comment]
> Hi @SXQ233, can you attach the failed `.pb` file to this issue?

Please download the failed .pb file from the link below:
[https://github.com/SXQ233/pb/blob/master/weights.pb](url)

**Comment by WanliZhong** [comment]
@SXQ233 Your link does not contain the model file. Can you provide another one?

**Comment by SXQ233** [comment]
> @SXQ233 Your link does not contain the model file. Can you provide another one?

Do not click the link directly. Please copy the link and paste it into the browser to open it.

**Comment by WanliZhong** [comment]
@SXQ233 I guess there is a problem in your model. The `Pack` node has 3 inputs, but two of them have no key to identify the meaning of the inputs.
![image](https://user-images.githubusercontent.com/49567307/210054797-2bc8849b-31b6-4728-92eb-e89bf09e39b9.png)

**Comment by WanliZhong** [comment]
@SXQ233 I'm not sure if the `Pack` node is right. Two of three inputs are const value without key to identify them.
![image](https://user-images.githubusercontent.com/49567307/210056542-b59f3f40-d9b5-46e7-b4d8-e8f7d1749908.png)

**Comment by SXQ233** [comment]
> @SXQ233 I'm not sure if the `Pack` node is right. Two of three inputs are const value without key to identify them. ![image](https://user-images.githubusercontent.com/49567307/210056542-b59f3f40-d9b5-46e7-b4d8-e8f7d1749908.png)

The program can run normally in python. Is there a problem when h5 is converted to pb?

> import tensorflow as tf
> from tensorflow.python.framework.convert_to_constants import convert_variables_to_constants_v2
> from tensorflow.keras.models import load_model
> 
> def h5_to_pb(model, pb_model_name):
>     model.summary()
>     full_model = tf.function(lambda Input: model(Input))
>     full_model = full_model.get_concrete_function(tf.TensorSpec(model.inputs[0].shape, model.inputs[0].dtype))
> 
>     # Get frozen ConcreteFunction
>     frozen_func = convert_variables_to_constants_v2(full_model)
>     frozen_func.graph.as_graph_def()
> 
>     layers = [op.name for op in frozen_func.graph.get_operations()]
>     print("-" * 50)
>     print("Frozen model layers: ")
>     for layer in layers:
>         print(layer)
> 
>     print("-" * 50)
>     print("Frozen model inputs: ")
>     print(frozen_func.inputs)
>     print("Frozen model outputs: ")
>     print(frozen_func.outputs)
> 
>     # Save frozen graph from frozen ConcreteFunction to hard drive
>     tf.io.write_graph(graph_or_graph_def=frozen_func.graph,
>                       logdir="./",
>                       name=f"{pb_model_name}.pb",
>                       as_text=False)
>     tf.io.write_graph(graph_or_graph_def=frozen_func.graph,
>                       logdir="./",
>                       name=f"{pb_model_name}.pbtxt",
>                       as_text=True)
> 
> h5_model_path = "logs/last.h5"
> pb_model_name = "logs/weights"
> 
> net_model = load_model(h5_model_path)
> h5_to_pb(net_model, pb_model_name)

**Comment by SXQ233** [comment]
@WanliZhong  The program can run normally in python. Is there a problem when h5 is converted to pb?

    # h5topb  
    import tensorflow as tf
    from tensorflow.python.framework.convert_to_constants import convert_variables_to_constants_v2
    from tensorflow.keras.models import load_model

    def h5_to_pb(model, pb_model_name):

        model.summary()
        full_model = tf.function(lambda Input: model(Input))
        full_model = full_model.get_concrete_function(tf.TensorSpec(model.inputs[0].shape, model.inputs[0].dtype))

        # Get frozen ConcreteFunction
        frozen_func = convert_variables_to_constants_v2(full_model)
        frozen_func.graph.as_graph_def()

        layers = [op.name for op in frozen_func.graph.get_operations()]
        print("-" * 50)
        print("Frozen model layers: ")
        for layer in layers:
            print(layer)

        print("-" * 50)
        print("Frozen model inputs: ")
        print(frozen_func.inputs)
        print("Frozen model outputs: ")
        print(frozen_func.outputs)

        # Save frozen graph from frozen ConcreteFunction to hard drive
        tf.io.write_graph(graph_or_graph_def=frozen_func.graph,
                      logdir="./",
                      name=f"{pb_model_name}.pb",
                      as_text=False)
        tf.io.write_graph(graph_or_graph_def=frozen_func.graph,
                      logdir="./",
                      name=f"{pb_model_name}.pbtxt",
                      as_text=True)
    h5_model_path = "logs/last.h5"
    pb_model_name = "logs/weights"

    net_model = load_model(h5_model_path)
    h5_to_pb(net_model, pb_model_name)
